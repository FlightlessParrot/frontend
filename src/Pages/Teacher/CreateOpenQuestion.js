import { Checkbox, Flex, FormControl, FormErrorMessage, FormLabel, Image, Input, Text } from "@chakra-ui/react";
import { Form, Link, useActionData } from "react-router-dom";
import { useEffect, useRef, useState } from "react";
import useShowToast from "../../hooks/useShowToast";
import Title from "../../Components/Title";

import PresentImage from "../../Components/Image/PresentImage";
import ImageForm from "../../Components/Image/ImageForm";

export default function CreateOpenQuestion() {
  const data=useActionData()
  const toast=useShowToast()
 
  const [question, setQuestion]=useState('')
  const [dirty, setDirty]=useState({
    question: false
  })
  const [check, setCheck]=useState(false)
  const [controller, setController]=useState(false)
  const fileRef=useRef(null)

  const keyRef=useRef(1)

useEffect(
  ()=>{
    console.log(controller)
    if(controller)
    {
    if(data?.questionId )
    { 
      toast(
        {
          title: 'Udało się',
          description: 'Pomyślnie dodano pytanie do testu.',
          status: 'success'
        } 
      )
      setQuestion('');
      keyRef.current++
      setCheck(false)
    }
    if(data!=null && !data?.questionId)
    { 
      toast(
        {
          title: 'Błąd',
          description: 'Nie udało się dodać pytania do testu.',
          status: 'error'
        }
      )
    }
    data && setController(false)
  }
  },[data,toast,setQuestion, controller, controller]
)


const questionIsInvalid=question==='' && dirty.question

  return (
   <Form key={keyRef.current} className="flex flex-col gap-8" method="post">
    <Title title='Dodaj otwarte pytania do egzaminu'/>
    <Flex direction={'column'} gap={20} p={[2,4,4,8,16]}>
      <ImageForm fileRef={fileRef} check={check} setCheck={setCheck} isValid/>
    <PresentImage fileRef={fileRef} />
    <FormControl isInvalid={questionIsInvalid}>
        <FormLabel>Dodaj pytanie</FormLabel>
        <Text color={"gray.500"} size='xs'>Pytanie nie może zawierać więcej niż 250 znaków</Text>
        <Input value={question} name='question' onChange={e=>{setDirty(s=>({...s,question: true}));setQuestion(e.target.value)}} required maxLength={250} placeholder="Czy pantofelek ma nibynóżkę?"></Input>
        <FormErrorMessage>Pytanie nie może być puste.</FormErrorMessage>
    </FormControl>
    <input readOnly className="hidden" name='type' value='open' />
    <Flex justify={'end'}><button className='action-button' onClick={(e)=>{
      if(questionIsInvalid ){ e.preventDefault()}
      else {setController(true)}
    }}>
      Dalej
      </button> </Flex>
    <Flex justify={'end'}><Link to='/user/team/egzam' className='action-button'>Zakończ przygotowanie egzaminu</Link></Flex>
    
    </Flex>
   </Form>
  )
}