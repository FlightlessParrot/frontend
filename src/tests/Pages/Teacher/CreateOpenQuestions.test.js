
import renderer from 'react-test-renderer'
import CreateOpenQuestion from '../../../Pages/Teacher/CreateOpenQuestion'
import { act, fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
global.URL.createObjectURL=jest.fn()
jest.mock('react-router-dom', ()=>{
    const router=jest.requireActual('react-router-dom');
    return {
        __esModule: true,
        ...router,
        useActionData: ()=>({}),
        Form: ({children})=><form>{children}</form>,
        Link: ({children})=><a>{children}</a>
    }
})
describe('test Create Open Questions Component',()=>{
    it('test component render correctly',()=>{
        const element=renderer.create(<CreateOpenQuestion />).toJSON()
        expect(element).toMatchSnapshot();
    })

    it('test user can write question name',()=>{
        render(<CreateOpenQuestion />)

        const question =screen.getByPlaceholderText(/pantofelek/)

        act(
            ()=>{
                userEvent.type(question,'text')
            }
        )

        expect(question).toHaveValue('text');
    })

    it('test errors appears',()=>{
        render(<CreateOpenQuestion />)
        const textFile = new File(['file'], 'file.txt', {type: 'txt'})
        const question =screen.getByPlaceholderText(/pantofelek/)
        const file=screen.getByTestId('fileInput')

        act(
            ()=>{
                fireEvent.change(question)
                userEvent.upload(file, textFile)
            }
        )
        const questionError=screen.getByText(/Pytanie nie może/)
        const fileError=screen.getByText('Format zdjęcia jest niewłaściwy.')

        expect(questionError).toBeInTheDocument()
        expect(fileError).toBeInTheDocument()

    })

    it('test errors dissapear when user gives correct data', ()=>
    {
        render(<CreateOpenQuestion />)
        const textFile =new File(['file'], 'file.txt', {type: 'txt'})
        const jpgFile= new File(['file'], 'file.jpg', {type: 'image/jpg'})
        const question =screen.getByPlaceholderText(/pantofelek/)
        const file=screen.getByTestId('fileInput')

        act(
            ()=>{
                fireEvent.change(question)
                userEvent.upload(file, textFile)
                
            }
        )
        const questionError=screen.queryByText(/Pytanie nie może być puste./)
        const fileError=screen.queryByText('Format zdjęcia jest niewłaściwy.')
        act(
            ()=>{
                userEvent.upload(file, jpgFile)
                userEvent.type(question,'question')
            }
        )
        expect(questionError).not.toBeInTheDocument()
        expect(fileError).not.toBeInTheDocument()
    })
})