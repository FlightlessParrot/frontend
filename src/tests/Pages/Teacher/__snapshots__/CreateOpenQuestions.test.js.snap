// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Create Open Questions Component test component render correctly 1`] = `
<form>
  <h1
    className="inline-block text-center p-4 md:p-8 px-12 border-b-4 article-title-font border-b-baltic "
    data-testid="Title-h1"
  >
    Dodaj otwarte pytania do egzaminu
  </h1>
  <div
    className="css-18y4t9"
  >
    <div
      className="chakra-form-control css-0"
      role="group"
    >
      <label
        className="chakra-form__label flex items-end p-4 cursor-pointer action-button gap-10 css-uvho5g"
        htmlFor="input"
        id="field-:r0:-label"
      >
        <label
          className="chakra-checkbox css-1o501kh"
          onClick={[Function]}
          readOnly={true}
        >
          <input
            aria-disabled={false}
            aria-invalid={false}
            checked={false}
            className="chakra-checkbox__input"
            disabled={false}
            id="field-:r0:"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            readOnly={true}
            required={false}
            style={
              Object {
                "border": "0",
                "clip": "rect(0, 0, 0, 0)",
                "height": "1px",
                "margin": "-1px",
                "overflow": "hidden",
                "padding": "0",
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "1px",
              }
            }
            type="checkbox"
          />
          <span
            aria-hidden={true}
            className="chakra-checkbox__control css-dnty2r"
            data-readonly=""
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
          />
        </label>
         Dodaj zdjęcie
      </label>
      <p
        className="chakra-text css-ax9bdp"
      >
        Dopuszczalne formaty to jpg, png oraz webp
      </p>
      <input
        accept=".jpg,.png,.webp"
        className="hidden"
        data-testid="fileInput"
        id="input"
        name="image"
        onChange={[Function]}
        type="file"
      />
    </div>
    <div
      className="chakra-form-control css-0"
      role="group"
    >
      <label
        className="chakra-form__label css-uvho5g"
        htmlFor="field-:r1:"
        id="field-:r1:-label"
      >
        Dodaj pytanie
      </label>
      <p
        className="chakra-text css-ax9bdp"
      >
        Pytanie nie może zawierać więcej niż 250 znaków
      </p>
      <input
        aria-required={true}
        className="chakra-input css-0"
        disabled={false}
        id="field-:r1:"
        maxLength={250}
        name="question"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Czy pantofelek ma nibynóżkę?"
        readOnly={false}
        required={true}
        value=""
      />
    </div>
    <div
      className="css-ckyhk3"
    >
      <button
        className="action-button"
        onClick={[Function]}
      >
        Dalej
      </button>
       
    </div>
    <div
      className="css-ckyhk3"
    >
      <a>
        Zakończ przygotowanie egzaminu
      </a>
    </div>
  </div>
</form>
`;
